File: 백준/Silver/12761. 돌다리/돌다리.py
import sys
input = sys.stdin.readline
from collections import deque

a, b, n, m = map(int, input().split())

q = deque()
visited = [0] * (100001)

def BFS():
    q.append(n)
    visited[n] = 1
    while q:
        x = q.popleft()
        dx = [x + 1, x - 1, x + a, x - a, x + b, x - b, x * a, x * b]
        for i in range(8):
            y = dx[i]
            if 0 <= y <= 100000 and not visited[y]:
                visited[y] = visited[x] + 1
                q.append(y)
            if y == m:
                return

BFS()
print(max(visited) - 1)

File: 백준/Silver/12761. 돌다리/돌다리.js
const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./example.txt")
  .toString()
  .trim()
  .split(" ")
  .map(Number);

const [a, b, n, m] = input;

const visited = Array(100001).fill(0);

function bfs() {
  q = [n];
  visited[n] = 1;
  while (q.length > 0) {
    const x = q.shift();
    const dx = [x - 1, x + 1, x - a, x + a, x - b, x + b, a * x, b * x];
    for (const v of dx) {
      if (!visited[v] && v >= 0 && v <= 100000) {
        visited[v] = visited[x] + 1;
        q.push(v);
      }
      if (v === m) return;
    }
  }
}
bfs();
console.log(Math.max(...visited) - 1);
